File: 16-defer/main.go
package main

import "fmt"

// 模拟资源
type Logger struct {
	Name string
}

func NewLogger(name string) *Logger {
	fmt.Printf("[%s] opened\n", name)
	return &Logger{Name: name}
}

func (l *Logger) Write(msg string) {
	fmt.Printf("[%s] %s\n", l.Name, msg)
}

func (l *Logger) Close() {
	fmt.Printf("[%s] closed\n", l.Name)
}

// 1. 基本 defer
func basicDefer() {
	fmt.Println("--- basicDefer ---")
	logger := NewLogger("basic")
	defer logger.Close() // 函数返回时关闭

	logger.Write("doing work...")
	logger.Write("more work...")
}

// 2. 多个 defer 按 LIFO 顺序执行
func multipleDefers() {
	fmt.Println("\n--- multipleDefers ---")
	defer fmt.Println("first defer (executes last)")
	defer fmt.Println("second defer (executes second)")
	defer fmt.Println("third defer (executes first)")
	fmt.Println("main body")
}

// 3. Go 的 defer 是函数级的（重要区别！）
func functionLevelDefer() {
	fmt.Println("\n--- functionLevelDefer ---")

	// 危险！所有 defer 都在函数返回时才执行
	// 如果循环次数多，会积累大量未释放的资源
	for i := 0; i < 3; i++ {
		logger := NewLogger(fmt.Sprintf("loop-%d", i))
		defer logger.Close() // 在函数返回时才执行，不是循环结束时！
		logger.Write("iteration work")
	}
	fmt.Println("After loop (defers haven't run yet!)")
	// 所有 Close 在此之后执行，顺序是 loop-2, loop-1, loop-0
}

// 4. 解决方案：用闭包创建新的函数作用域
func scopedDeferWorkaround() {
	fmt.Println("\n--- scopedDeferWorkaround ---")

	for i := 0; i < 3; i++ {
		func() {
			logger := NewLogger(fmt.Sprintf("scoped-%d", i))
			defer logger.Close() // 在匿名函数返回时执行
			logger.Write("iteration work")
		}()
	}
	fmt.Println("After loop (all loggers already closed)")
}

// 5. defer 参数在声明时求值
func deferEvaluation() {
	fmt.Println("\n--- deferEvaluation ---")

	x := 10
	defer fmt.Println("deferred x =", x) // x=10 在此时就确定了
	x = 20
	fmt.Println("current x =", x) // 20
	// defer 输出的是 10，不是 20！
}

// 6. defer 与 panic/recover
func safeOperation() {
	fmt.Println("\n--- safeOperation ---")

	defer func() {
		if r := recover(); r != nil {
			fmt.Println("Recovered from panic:", r)
		}
	}()

	fmt.Println("Before panic")
	panic("something went wrong")
	// 下面的代码不会执行
}

// 7. Go 没有 errdefer，需要手动处理
func riskyOperation(shouldFail bool) (string, error) {
	fmt.Printf("\n--- riskyOperation(fail=%v) ---\n", shouldFail)

	logger := NewLogger("risky")
	// Go 没有 errdefer，必须无条件 defer 或手动管理
	// 方式1：无条件关闭（即使成功也关闭）
	// defer logger.Close()

	// 方式2：根据错误状态决定（更灵活但更啰嗦）
	var err error
	defer func() {
		if err != nil {
			fmt.Println("Cleaning up due to error")
			logger.Close()
		}
	}()

	logger.Write("doing risky work")
	if shouldFail {
		err = fmt.Errorf("operation failed")
		return "", err
	}
	logger.Write("success!")
	return "result", nil
}

func main() {
	basicDefer()
	multipleDefers()
	functionLevelDefer()
	scopedDeferWorkaround()
	deferEvaluation()
	safeOperation()
	riskyOperation(true)
	riskyOperation(false)
}

File: 16-defer/main.zig
const std = @import("std");
const print = std.debug.print;

// 模拟资源：一个简单的"日志记录器"
const Logger = struct {
    name: []const u8,

    fn init(name: []const u8) Logger {
        print("[{s}] opened\n", .{name});
        return Logger{ .name = name };
    }

    fn write(self: Logger, msg: []const u8) void {
        print("[{s}] {s}\n", .{ self.name, msg });
    }

    fn close(self: Logger) void {
        print("[{s}] closed\n", .{self.name});
    }
};

// 1. 基本 defer 使用
fn basicDefer() void {
    print("--- basicDefer ---\n", .{});
    const logger = Logger.init("basic");
    defer logger.close(); // 函数返回前关闭

    logger.write("doing work...");
    logger.write("more work...");
    // logger.close() 会在此自动执行
}

// 2. 多个 defer 按逆序执行
fn multipleDefers() void {
    print("\n--- multipleDefers ---\n", .{});
    defer print("first defer (executes last)\n", .{});
    defer print("second defer (executes second)\n", .{});
    defer print("third defer (executes first)\n", .{});
    print("main body\n", .{});
}

// 3. 作用域级 defer（与 Go 的关键区别！）
fn scopedDefer() void {
    print("\n--- scopedDefer ---\n", .{});

    var i: usize = 0;
    while (i < 3) : (i += 1) {
        // 这个 defer 在每次循环迭代结束时执行
        // （不是在函数返回时！）
        const logger = Logger.init("loop");
        defer logger.close();

        logger.write("iteration work");
        // logger.close() 在每次迭代结束时执行
    }
    print("After loop\n", .{});
}

// 4. errdefer - 仅在错误路径执行（Zig 独创！）
fn riskyOperation(should_fail: bool) ![]const u8 {
    print("\n--- riskyOperation(fail={}) ---\n", .{should_fail});

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const buf = try allocator.alloc(u8, 32);
    // 仅在返回错误时释放 buf
    errdefer {
        print("errdefer: freeing buffer due to error!\n", .{});
        allocator.free(buf);
    }

    if (should_fail) {
        return error.OperationFailed;
        // errdefer 会执行，释放 buf
    }

    @memcpy(buf[0..5], "hello");
    // 成功路径：errdefer 不执行，调用者负责释放
    return buf[0..5];
}

// 5. defer 在块作用域中
fn blockScopeDemo() void {
    print("\n--- blockScopeDemo ---\n", .{});

    print("before block\n", .{});
    {
        defer print("defer in inner block\n", .{});
        print("inside block\n", .{});
    } // <-- 内部 defer 在此执行
    print("after block\n", .{});
}

pub fn main() void {
    // 1. 基本 defer
    basicDefer();

    // 2. 多个 defer 逆序执行
    multipleDefers();

    // 3. 作用域级 defer（循环安全）
    scopedDefer();

    // 4. errdefer 演示
    // 失败路径：errdefer 执行
    _ = riskyOperation(true) catch |err| {
        print("Caught error: {}\n", .{err});
    };

    // 成功路径：errdefer 不执行
    if (riskyOperation(false)) |result| {
        print("Success: {s}\n", .{result});
        // 注意：实际代码中应该释放 result 的内存
    } else |err| {
        print("Error: {}\n", .{err});
    }

    // 5. 块作用域 defer
    blockScopeDemo();
}
